homeassistant/components/rflink.py

```diff
@@ -0,0 +1,269 @@
+"""Support for Rflink components.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/rflink/
+
+Technical overview:
+
+The Rflink gateway is a USB serial device (Arduino with Rflink firwmare)
+connected to a 433Mhz transceiver module.
+
+The the `rflink` Python module a asyncio transport/protocol is setup that
+fires an callback for every (valid/supported) packet received by the Rflink
+gateway.
+
+This component uses this callback to distribute 'rflink packet events' over
+the HASS bus which can be subscribed to by entities/platform implementations.
+
+The platform implementions take care of creating new devices (if enabled) for
+unsees incoming packet id's.
+
+Device Entities take care of matching to the packet id, interpreting and
+performing actions based on the packet contents. Common entitiy logic is
+maintained in this file.
+
+"""
+import asyncio
+import logging
+
+from homeassistant.const import (
+    ATTR_ENTITY_ID, EVENT_HOMEASSISTANT_STOP, STATE_UNKNOWN)
+from homeassistant.helpers.entity import Entity
+
+REQUIREMENTS = ['rflink==0.0.18']
+
+DOMAIN = 'rflink'
+
+RFLINK_EVENT = {
+    'light': 'rflink_switch_event_received',
+    'sensor': 'rflink_sensor_event_received',
+    'switch': 'rflink_switch_event_received',
+    'send_command': 'rflink_send_command',
+}
+
+ATTR_EVENT = 'event'
+ATTR_COMMAND = 'command'
+
+_LOGGER = logging.getLogger(__name__)
+
+KNOWN_DEVICE_IDS = []
+
+
+def identify_event_type(event):
+    """Look at event to determine type of device."""
+    if 'command' in event:
+        return 'light'
+    elif 'sensor' in event:
+        return 'sensor'
+    else:
+        return 'unknown'
+
+
+@asyncio.coroutine
+def async_setup(hass, config):
+    """Setup the Rflink component."""
+    from rflink.protocol import create_rflink_connection
+
+    # make sure no known devices are left (mostly during tests)
+    del KNOWN_DEVICE_IDS[:]
+
+    def event_callback(event):
+        """Handle incoming rflink events.
+
+        Rflink events arrive as dictionaries of varying content
+        depending on their type. Identify the events and distribute
+        accordingly.
+
+        """
+        event_type = identify_event_type(event)
+        _LOGGER.info('event type %s', event_type)
+
+        # fire bus event for event type
+        if event_type in RFLINK_EVENT:
+            hass.bus.fire(RFLINK_EVENT[event_type], {ATTR_EVENT: event})
```

Comment:
Sorry, many was a bit of an overstatement indeed. Two components that do seem to use this pattern are:
- https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/zigbee.py#L128-L135
- https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/switch/pilight.py#L90-L91

Others just output to the bus for user to subscribe to but don't use it internally indeed, did not see that before, sorry. 

I will refactor to not use the bus for propagating events internally. 

I think I will keep firing some of the events to allow the user to subscribe to things like button presses instead of just state changes.